### PR TITLE
build and patch from upstream

### PR DESCRIPTION
**Pull request by ab77**
patch and build Debian kernels from [upstream](https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/jammy/), relying on [renovate regex parser to keep the submodule ref up to date](https://github.com/balena-io/renovate-config/pull/389) and publish as github releases (*.deb).

Eventually replacing parts of the following:
https://github.com/balena-io/remote-builders/blob/master/equinix_metal_devices.tf#L9-L10
https://github.com/balena-io/remote-builders/blob/master/ubuntu/setup.sh#L19-L25

(WIP) https://github.com/balena-io/remote-builders/compare/master...ab77/operational?quick_pull=1

### CODE REVIEW

File: .github/actions/publish/action.yml [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        trap 'cleanup' EXIT
+
+        pushd build/kernel-src
+        for kc in ${ANNOTATE_IGNORE}; do
+            sed -i "s/${kc}/#${kc}/g" debian.master/config/annotations
+        done
+        popd
+
+      env:
+        ANNOTATE_IGNORE: 'CONFIG_DEBUG_INFO_DWARF_TOOLCHAIN_DEFAULT CONFIG_DEBUG_INFO_DWARF5'
```
Comment (jakogut) [inline .github/actions/publish/action.yml:73]: I like sticking the annotations to remove in the env var, nice.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    rpmbuild -bp SPECS/kernel.spec
+
+    pushd BUILD/kernel-*
+
+    kernel_tree=$(find . -maxdepth 1 -name linux-*.amzn2.aarch64 -type d)
+
+    cp -a ${kernel_tree}{,.org}
+
+
+### patch
```
Comment (jakogut) [inline README.md:57]: This patch may not be necessary with the Ubuntu kernel sources. See [this patch](https://github.com/balena-os/balena-generic/blob/master/layers/meta-balena-generic/recipes-kernel/linux/files/generic-aarch64/0001-UBUNTU-SAUCE-no-up-add-compat_uts_machine-kernel-com.patch) applied in balenaOS from the Ubuntu kernel. It adds the `compat_uts_machine` kernel cmdline parameter that allows for overriding `uts_machine` with an arbitrary string at boot.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+  * Instead we map 3.x to 2.6.40+x, so e.g. 3.0 would be 2.6.40
+- * And we map 4.x and later versions to 2.6.60+x, so 4.0/5.0/6.0/... would be
+- * 2.6.60.
++ * And we map 4.x to 2.6.60+x, so 4.0 would be 2.6.60.
+  */
+ static int override_release(char __user *release, size_t len)
+ {
+    int ret = 0;
+
++   strncpy(compat_uts_machine, "armv7l", __OLD_UTS_LEN);
```
Comment (jakogut) [inline README.md:102]: Here we override with the static string `armv7l` if `PERLINUX32` is enabled, and below, we override with `armv6l` if `UNAME26` is enabled. I suppose this has the benefit of switching between three values of `uts_machine` at runtime, but it's hacky, and doesn't allow arbitrary machine strings to be set at boot.

**Review by jakogut — COMMENTED** [review]
Looks good to me. Minor comment on the necessity of the balena specific `uts_machine` hacky patch. If we're able to dedicate builders to a single architecture each, we can probably drop this patch altogether in favor of the upstream Ubuntu solution with the `compat_uts_machine` cmdline parameter set at boot. If that's not feasible, the trade off with our hacky patch allowing three unique values to be used at runtime for `uts_machine` is fine by me.

File: README.md (hunk 3 above)
Comment (ab77) [reply README.md:102]: @jakogut the current architecture requires x3 Docker sockets, each reporting one of these architectures:

* armv6l
* armv7l
* aarch64

**Comment by ab77** [comment]
> Looks good to me. Minor comment on the necessity of the balena specific `uts_machine` hacky patch. If we're able to dedicate builders to a single architecture each, we can probably drop this patch altogether in favor of the upstream Ubuntu solution with the `compat_uts_machine` cmdline parameter set at boot. If that's not feasible, the trade off with our hacky patch allowing three unique values to be used at runtime for `uts_machine` is fine by me.

We could but it costs $$$, so we don't.

**Comment by ab77** [comment]
lgtm

**Pull request merged**